File: docs/sphinx/examples/cpp/providers/nvcf_sample.cpp
// Compile and run with:
// ```
// nvq++ --target nvcf nvcf_sample.cpp -o out.x
// ./out.x
// ```
// Assumes a valid NVCF API key and function ID have been stored in environment
// variables or `~/.nvcf_config` file. Alternatively, they can be set in the
// command line like below.
// ```
// nvq++ --target nvcf --nvcf-api-key <YOUR API KEY> --nvcf-function-id \ 
// <NVCF function Id> nvcf_sample.cpp -o out.x
// ./out.x
// ```
// Please refer to the documentations for information about how to attain NVCF
// information.

#include <cudaq.h>
#include <iostream>

// Define a simple quantum kernel to execute on NVCF.
struct ghz {
  // Maximally entangled state between 25 qubits.
  auto operator()() __qpu__ {
    constexpr int NUM_QUBITS = 25;
    cudaq::qvector q(NUM_QUBITS);
    h(q[0]);
    for (int i = 0; i < NUM_QUBITS - 1; i++) {
      x<cudaq::ctrl>(q[i], q[i + 1]);
    }
    auto result = mz(q);
  }
};

int main() {
  // Submit to NVCF asynchronously (e.g., continue executing
  // code in the file until the job has been returned).
  auto async_counts_handle = cudaq::sample_async(ghz{});
  // ... classical code to execute in the meantime ...
  std::cout << "Waiting for NVCF result...\n";

  // Calling .get() on the handle to synchronize the result.
  auto async_counts = async_counts_handle.get();
  async_counts.dump();

  // OR: Submit to NVCF synchronously (e.g., wait for the job
  // result to be returned before proceeding).
  auto counts = cudaq::sample(ghz{});
  counts.dump();
}

File: docs/sphinx/examples/python/providers/nvcf_sample.py
import cudaq

# This example assumes the NVCF API key and Function Id have been set in the `~/.nvcf_config` file/environment variables.
# If not, you can set the API Key and Function ID environment variables in the Python script with:
# ```
# os.environ["NVCF_API_KEY"] = "<YOUR NVCF API KEY>"`
# os.environ["NVCF_FUNCTION_ID"] = "<YOUR NVCF FUNCTION ID>"
# ```
# Alternatively, the `api_key` and `function_id` values can be passed to the target directly,
# ```
# cudaq.set_target("nvcf",
#                 backend="tensornet",
#                 api_key="<YOUR NVCF API KEY>"
#                 function_id="<YOUR NVCF FUNCTION ID>")
# ```
cudaq.set_target("nvcf", backend="tensornet")

num_qubits = 50


@cudaq.kernel(jit=True)
def ghz(num_qubits: int):
    qubits = cudaq.qvector(num_qubits)
    # Place qubits in GHZ state.
    h(qubits[0])
    for i in range(num_qubits - 1):
        x.ctrl(qubits[i], qubits[i + 1])
    # Measure.
    mz(qubits)


counts = cudaq.sample(ghz, num_qubits, shots_count=100)
print(counts)
